enhancements/baremetal/debuggability-of-baremetal-ipi-deployment.md

```diff
@@ -0,0 +1,246 @@
+we do not provide useful information to the user about what to do.
+
+#### Bootstrap Failures
+
+The bootstrap runs a couple of baremetal-specific services, including
+Ironic as well as a utility that populates introspection data for
+the control plane.
+
+Bootstrap typically fails for baremetal when we can't download the
+machine-os image into our local HTTP cache.  Less common, but still
+sometimes seen are that services such as dnsmasq, mariadb, ironic-api,
+ironic-conductor, or ironic-inspector fail.
+
+Failures on bootstrap services rarely result in any indication to the
+user that something went wrong other than that there was a timeout.
+
+The installer has a feature for log gathering on bootstrap failure that
+does not work on baremetal. This should be the first priority, but even
+in this case a user still needs to look into an archive containing many
+logs to identify a failure.
+
+Ideally there would be some mechanism to identify and extract useful
+information and display it to the user.
+
+#### Bare Metal Host Provisioning
+
+Whether the control plane or worker nodes, provisioning of bare metal
+hosts can fail in the same ways, although the communication path to
+provide feedback is different in each case. For the control plane,
+information about failure is presented to the user via terraform. For
+workers, it would be through information on the `BareMetalHost`
+resource, and the baremetal-operator logs.
+
+Provisioning can fail in many ways. The most difficult to troubleshoot
+are simply when we fail to hear back from a host. Buggy UEFI firmware
+may prevent PXE, a kernel could panic, or even a network cable may be
+unplugged. In these cases, we should inform the user what little
+information Ironic was able to discern, but also provide a suggestion
+that the most effective way to troubleshoot the problem is examination
+of the console of the host.
+
+An infrequent, but possible outcome of deployment to bare metal hosts,
+is that Ironic is successful in cleaning, inspecting, and deploying a
+host. After Ironic lays down an image on disk and reboots, Ironic marks
+the host ‘active’.  However, when the host boots again it’s possible that
+there’s a catastrophic problem such as a kernel panic or fail to
+configure with ignition. From Ironic's perspective, it's done it's duty,
+and is unaware the host failed to come up. The feedback to the user is
+only that there was a timeout.
+
+#### Operator Deployment
+
+Operator deployment failures are rarely platform-specific, although
+there is one case that should be addressed. When worker deployment
+fails, possibly due to provisioning issues like those described above, a
+variety of operators may report failures such as ingress, console, and
+others that cannot run on the control plane.
+
+When this happens, the installer times out, reports to the user a large
+number of operators failed to roll out, and no useful context about what
+to do or why the operators failed.
```

Comment:
that's on the operator owners to make sure the errors are clear. Think about how it is not only installer that is the consumer of these message, but also admins during upgrades.

So personally the goal should be to ensure that each operator is responsible for using clear error messages in the status.

Reply:
I agree, it's just on day 1 the worker deployment failure seems to be special to me. It causes a lot of noise as a bunch of operators start reporting error messages that make it hard to point to a root cause unless you've seen the problem before. I don't think machine-api operator even reports anything useful when this happens, but if it did, it'd get lost in mix of the many other failing operators.

Reply:
If we take ingress/console operator as an example - if the worker fails, they will be in error - but, from my experience of installing openshift for first few times - the user will have no idea that this is the reason. He will just see that those operators are down.
What is possible maybe to do is to have some kind of 'validators' - either from the installer binary or as an operator - that can analyze logs or cluster runtime state (with minimal requirement for cluster functionality - such as passwordless ssh between nodes) that can look into the state of the cluster and explain the user what went wrong. If we provide an infra for writing those validators, then operators owners / qe / intergration team will be able to enhance those once they rn into an issue that it was hard to analyze.

Reply:
If they asked for a certain number of works and they didn't get them that seems reasonable to have a special error for that.

I also think some generic orientation regarding how to investigate operators failing may help as well. They'll need to learn that skill eventually no matter what. So documenting how to look at an Operator's status and referencing that seems to be something worth doing no matter what.